src/java/org/apache/cassandra/db/streaming/ComponentInfo.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.streaming;
+
+import java.io.IOException;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.io.IVersionedSerializer;
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+
+public class ComponentInfo
+{
+    final Component.Type type;
+    final long length;
+
+    public ComponentInfo(Component.Type type, long length)
+    {
+        assert length >= 0 : "Component length cannot be negative";
+        this.type = type;
+        this.length = length;
+    }
+
+    @Override
+    public String toString()
+    {
+        return "ComponentInfo{" +
+               "type=" + type +
+               ", length=" + length +
+               '}';
+    }
+
+    public boolean equals(Object o)
```

Comment:
It's generally considered to be a bad practice to implement `equals()` and `hashCode()` unless that class is stored in a set or a map - or an upstream implementation of such. Otherwise it's just confusing boilerplate (confusing because it implies that the class is used in ways it clearly isn't).

In this case, there is a `List<ComponentInfo>` field in `CassandraStreamHeader`, which has an `equals()`/`hashCode()` implementation, which on the surface justifies these. But those, in turn, are actually dead code. So what we should do is remove the implementations of `equals()` and `hashCode()` here, and do the same in `CassandraStreamHeader`, being good citizens.

Reply:
I shouldn't have said "it's generally considered" here, as apparently it's not quite, as @aweisberg pointed out. It's a subjective preference, and what I'd do, but feel free to disregard this particular comment.

Reply:
`CassandraStreamHeader`'s `equals()` and `hashCode()` method are not dead code. They're used in `CassandraStreamHeaderTest::serializerTest` which eventually ends up in [`SerializationUtils::assertSerializationCycle`](https://github.com/apache/cassandra/blob/9714a7c817b64a3358f69e536535c756c5c6df48/test/unit/org/apache/cassandra/serializers/SerializationUtils.java#L60). This is what originally forced me to implement those methods. The only other choice was to test the equality for `ComponentInfo` in `CassandraStreamHeader::equal()` which would have been a bad choice.